src/ios-deploy/ios-deploy.m

```diff
@@ -1714,30 +1729,84 @@ void handle_device(AMDeviceRef device) {
         check_error(AMDeviceStopSession(device));
         check_error(AMDeviceDisconnect(device));
 
-        CFStringRef keys[] = { CFSTR("PackageType") };
-        CFStringRef values[] = { CFSTR("Developer") };
-        CFDictionaryRef options = CFDictionaryCreate(NULL, (const void **)&keys, (const void **)&values, 1, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
-
-        //assert(AMDeviceTransferApplication(afcFd, path, NULL, transfer_callback, NULL) == 0);
-        check_error(AMDeviceSecureTransferPath(0, device, url, options, transfer_callback, 0));
-        close(*afcFd);
-
+        bool incremental_install = app_deltas != NULL;
 
+        CFDictionaryRef options;
+        if (!incremental_install) {
+          CFStringRef keys[] = { CFSTR("PackageType") };
+          CFStringRef values[] = { CFSTR("Developer") };
+          options = CFDictionaryCreate(NULL, (const void **)&keys, (const void **)&values, 1, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+          //assert(AMDeviceTransferApplication(afcFd, path, NULL, transfer_callback, NULL) == 0);
+          check_error(AMDeviceSecureTransferPath(0, device, url, options, transfer_callback, 0));
+          close(*afcFd);
+        }
 
         AMDeviceConnect(device);
         assert(AMDeviceIsPaired(device));
         check_error(AMDeviceValidatePairing(device));
         check_error(AMDeviceStartSession(device));
 
-        // // NOTE: the secure version doesn't seem to require us to start the installation_proxy service
-        // // Although I can't find it right now, I in some code that the first param of AMDeviceSecureInstallApplication was a "dontStartInstallProxy"
-        // // implying this is done for us by iOS already
+        if (!incremental_install) {
+          // // NOTE: the secure version doesn't seem to require us to start the installation_proxy service
+          // // Although I can't find it right now, I in some code that the first param of AMDeviceSecureInstallApplication was a "dontStartInstallProxy"
+          // // implying this is done for us by iOS already
 
-        //service_conn_t installFd;
-        //assert(AMDeviceSecureStartService(device, CFSTR("com.apple.mobile.installation_proxy"), NULL, &installFd) == 0);
+          //service_conn_t installFd;
+          //assert(AMDeviceSecureStartService(device, CFSTR("com.apple.mobile.installation_proxy"), NULL, &installFd) == 0);
 
-        //mach_error_t result = AMDeviceInstallApplication(installFd, path, options, install_callback, NULL);
-        check_error(AMDeviceSecureInstallApplication(0, device, url, options, install_callback, 0));
+          //mach_error_t result = AMDeviceInstallApplication(installFd, path, options, install_callback, NULL);
+          check_error(AMDeviceSecureInstallApplication(0, device, url, options, install_callback, 0));
+        } else {
+          if (app_deltas == NULL) {
```

Comment:
With the current logic app_deltas can't be NULL here so this check can be dropped. (personally I think getting rid of `bool incremental_install = app_deltas != NULL;` and changing the checks for that bool to `app_deltas == NULL` everywhere would be easier to read.)

Reply:
Done.